### DISCUSSION TITLE
Failing API calls RTKQ Query make jest fail

### DISCUSSION THREAD

**Discussion by jeremiebardon** [mswjs/msw#1601] [Q&A] [answered]
Hello, since I think it's not an internal issue with MSW but a misunderstanding from my side I would like to kindly ask for help about my issue there!

I want to test an async function with a lazy query (RTK Query) in a custom hook. This function is exported for being use later. This function look like:

```
const authenticateUser = useCallback(
    async (token: string) => {
      try {
        await startAuthenticate(token, false).unwrap();
        onAuthSuccessCallback?.();
      } catch (err) {
        onAuthFailureCallback?.(err);
      }
    },
    [startAuthenticate, onAuthSuccessCallback, onAuthFailureCallback],
  );
```

Then this is my test:

```
server.use(
   rest.get(`${API_BASE_URL}api/v2.1/session`, (req, res, ctx) => {
       ctx.set('Authorization', 'Bearer token');
          return res(
            ctx.json({
              message: 'Could not be identified',
            }),
            ctx.status(403),
          );
        }),
      );

      const appStore = createStore({
        preloadedState: {
          session: sessionStateMock,
        },
      });

      const { result } = renderHook(useAuth, {
        initialProps: {
          onAuthFailureCallback: onAuthFailureCallbackMock,
          onAuthSuccessCallback: onAuthSuccessCallbackMock,
        },
        wrapper: props => <HookWrapper {...props} appStore={appStore} />,
      });

      await act(async () => {
        await result.current.authenticateUser('token');
      });

      expect.assertions(1);
      expect(onAuthFailureCallbackMock).toHaveBeenCalledWith({
        status: 403,
        data: { message: 'Could not be identified' },
      });
    });
```

The test is correctly asserted, if I'm changing the status or message he fail and tell me he expect the value shown above but the tests still fail and throw the API error like so:

```
thrown: Object {
      "error": Object {
        "data": Object {
          "message": "Could not be identified",
        },
        "status": 403,
      },
...rest
}

at _getError (node_modules/jest-circus/build/utils.js:424:18)
          at Array.map (<anonymous>)
```

And finally this is the configuration file:

```
import 'whatwg-fetch';
import '@testing-library/jest-native/extend-expect';

import { fetch, Headers, Request, Response } from 'cross-fetch';

import petalApi from '~api/index';
import { createStore } from '~store/index';

import { server } from './server.mocks';

global.fetch = fetch;
global.Headers = Headers;
global.Request = Request;
global.Response = Response;

const store = createStore({});

beforeAll(() => {
  server.listen({ onUnhandledRequest: 'error' });
});

afterEach(() => {
  server.resetHandlers();

  store.dispatch(petalApi.util.resetApiState());
});

afterAll(() => {
  server.close();
});
```

The configuration seems to work well on other assertion (On success), I'm guessing that my implementation is the problem, did you have an idea where this error can come from ?

**Comment by kettanaito** [comment, marked as answer]
Hi, @jeremiebardon. 

I think your approach on this is overall correct. I've spotted a few minor mistakes like calling `ctx.set()` as a side-effect (that won't do anything) instead of adding it to the `res()` composition:

```diff
-ctx.set('Authorization', 'Bearer token');
-return res(ctx.json({})

+ return res(ctx.set('Authorization', 'Bearer token'), ctx.json({})
```

> The test is correctly asserted, if I'm changing the status or message he fail and tell me he expect the value shown above but the tests still fail and throw the API error 

This sounds like a test structure issue. Take a closer look at how your custom hook handles rejections because to me this reads like it _throws_ on rejections, which propagates the exception to your test and fails it. 

A nitpick, but I think the callback is quite overcomplicated. You can achieve the same with this:

```ts
const authenticateUser = useCallback((token: string) => {
  return startAuthenticate(token, false).unwrap().then(
    onAuthSuccessCallback,
    onAuthFailureCallback
  )
}, [startAuthenticate, onAuthSuccessCallback, onAuthFailureCallback])
```

But you may disregard this suggestion because it's also about the error-handling behavior that you want to achieve. Your current implementation should swallow the errors. Mine above will forward them up unless you provide an error callback to handle them. This may or may not be what you're aiming for.

**Reply by jeremiebardon** [reply]
Spot on @kettanaito ! You were right about a mistake I've made on my API, in my RTK query function onQueryStarted I have not correctly catch the error now that is done it works correctly! My bad, thank you for your help ! :) 

Thank you for this library by the way, it's really awesome to build tests around "almost" real environment, help us a lot to ship our code with more confidence 🥇

**Reply by kettanaito** [reply]
Glad to hear you solved it!

Thank you for your kind words. If you like MSW and want it to improve and grow, please consider [Sponsoring it on GitHub](https://github.com/sponsors/mswjs). You can raise this conversation with your manager and provide a company sponsorship as an example. It would help the library a lot and we could meet the milestones faster that way. Thank you!
